rust/src/core.rs

```diff
@@ -163,6 +166,83 @@ impl FilesystemScriptPrep {
     }
 }
 
+/// Some Fedora/RHEL kernels ship .hmac files with absolute paths inside,
+/// which breaks when we relocate them into ostree/.  This function
+/// changes them to be relative.
+///
+/// Fixes are pending to ensure the kernels are doing this from the start,
+/// but we can't rely on that landing for a while. In addition, there are
+/// many 3rd-party kernels that have replicated the behavior in the ARK
+/// kernel.spec.
+fn verify_kernel_hmac_impl(moddir: &Dir) -> Result<()> {
+    use std::os::unix::ffi::OsStrExt;
+
+    // FIXME: in 2023
+    // This method is intentionally a misnomer because it should eventually
+    // be changed to the "sanity check" (below). It currently patches absolute
+    // paths to give kernel package maintainers time to update their .spec files.
+
+    const SEPARATOR: &[u8] = b"  ";
+
+    let hmac_path = ".vmlinuz.hmac";
+
+    let hmac_contents = if let Some(f) = moddir.open_optional(hmac_path)? {
+        let mut f = BufReader::new(f);
+        let mut s = Vec::new();
+        f.read_to_end(&mut s)?;
+        s
+    } else {
+        return Ok(());
+    };
+
+    if !hmac_contents.contains(&b'/') {
+        return Ok(());
+    }
+
+    let split_index = hmac_contents
+        .windows(SEPARATOR.len())
+        .position(|v| v == SEPARATOR)
+        .ok_or_else(|| {
+            anyhow::anyhow!(
+                "Missing path in .vmlinuz.hmac: {}",
+                String::from_utf8_lossy(hmac_contents.as_slice())
+            )
+        })?;
+
+    let (hmac, path) = hmac_contents.split_at(split_index + SEPARATOR.len());
+    let path = OsStr::from_bytes(path);
+
+    let file_name = Path::new(path).file_name().ok_or_else(|| {
```

Comment:
For OSTree based systems, it is a hard error to have non-UTF-8 paths.  In other bits of the code we've been using `camino::UTF8Path`.  This would make sense to switch to after `hmac_contents` is a `String.

Reply:
Thanks, I will keep this assumption in mind going forward.